backend/src/Database.hs

```diff
@@ -1,108 +1,111 @@
     PlayerId,
-    PlayerStats (..),
+    PlayerStatsTotal (..),
+    PlayerStatsYear (..),
     RatingDiff,
     Unique (..),
     currentYear,
     defaultPlayerStats,
-    rolloverPlayerStats,
   )
 import Prelude hiding (get, on)
 
-getPlayerByName :: (MonadIO m, MonadLogger m) => PlayerName -> SqlPersistT m (Maybe (Entity Player))
-getPlayerByName name = getBy $ UniquePlayerName name
+type DBAction m = ExceptT ServerError (SqlPersistT m)
 
-insertPlayerIfNotExists :: (MonadIO m, MonadLogger m) => PlayerName -> SqlPersistT m (Key Player)
-insertPlayerIfNotExists name =
-  getPlayerByName name >>= \case
+runDb :: DBAction (LoggingT IO) a -> AppM a
+runDb dbAction = do
+  env <- ask
+  result <- liftIO . runAppLogger env.logger . runSqlPool (runExceptT dbAction) $ env.dbPool
+  case result of
+    Left err -> throwError err
+    Right a -> pure a
+
+getPlayerByName :: (MonadIO m, MonadLogger m) => PlayerName -> DBAction m (Maybe (Entity Player))
+getPlayerByName = lift . getBy . UniquePlayerName
+
+insertPlayerIfNotExists :: (MonadIO m, MonadLogger m) => PlayerName -> DBAction m (Key Player)
+insertPlayerIfNotExists name = do
+  player <- getPlayerByName name
+  case player of
     Just (Entity playerKey _) -> pure playerKey
-    Nothing -> do
+    Nothing -> lift $ do
       logInfoN $ "Adding new player " <> name <> " to database."
       playerKey <- insert $ Player name Nothing
       year <- currentYear
-      insert_ $ defaultPlayerStats playerKey year
+      let (totalStats, yearStats) = defaultPlayerStats playerKey year
+      insert_ totalStats
+      insert_ yearStats
       pure playerKey
 
-getStats :: (MonadIO m, MonadLogger m) => PlayerId -> Year -> SqlPersistT m PlayerStats
-getStats pid year =
-  get (PlayerStatsKey pid (fromIntegral year)) >>= \case
-    Just stats -> pure stats
-    Nothing -> do
-      priorStats <- getMostRecentStats pid -- TODO Buggy sadness https://github.com/sirrus233/wotr-community-website/pull/31#discussion_r1897581825
-      let priorYear = priorStats.playerStatsYear
-      logInfoN $ "No stats for PID " <>: pid <> " in year " <>: year <> ". Rolling over from " <>: priorYear <> "."
-      let stats = rolloverPlayerStats pid year priorStats
-      insert_ stats
-      pure stats
-
-getMostRecentStats :: (MonadIO m, MonadLogger m) => PlayerId -> SqlPersistT m PlayerStats
-getMostRecentStats pid =
-  recentStats >>= \case
-    Just (Entity _ stats) -> pure stats
-    Nothing -> do
-      logErrorN $ "Missing stats for PID " <>: pid <> ". Inserting defaults."
-      year <- currentYear
-      let stats = defaultPlayerStats pid year
-      insert_ stats
-      pure stats
-  where
-    recentStats = selectOne $ do
-      stats <- from $ table @PlayerStats
-      where_ (stats ^. PlayerStatsPlayerId ==. val pid)
-      orderBy [desc (stats ^. PlayerStatsYear)]
-      pure stats
+getPlayerStats :: (MonadIO m, MonadLogger m) => PlayerId -> DBAction m (Maybe (Entity PlayerStatsTotal, Entity PlayerStatsYear))
```

Comment:
I decided to delete a bunch of the buggy year-stat handling stuff, and just not worry about it for now (for MVP). This just reads stats for the current year, and assumes they exist. We can figure out a longer-term plan later.